### PR TITLE
Skills - Implement Infantry man skill tree

### PR DESCRIPTION
**Pull request by cerebral-314**
- https://app.clickup.com/6631910/v/l/6acf6-1952
- ~~https://docs.google.com/spreadsheets/d/17mmOFfmvGePKPaiMfS4Xxzo5DbqLoE-tB_mW--oB3dU/edit#gid=944780773~~

### CODE REVIEW

File: game/functions/systems/skills/global/fn_skills_parseTreeCfg.sqf [hunk 1]
```diff
@@ -47,6 +47,7 @@ private _fnc_parseSkillTree = {
                 ["displayName", getText (_x >> "displayName")],
                 ["description", getText (_x >> "description")],
                 ["icon", getText (_x >> "icon")],
+                ["skillType", getNumber (_x >> "skillType")],
                 ["isActive", getNumber (_x >> "skillType") > 0],
```
Comment (veteran29) [inline game/functions/systems/skills/global/fn_skills_parseTreeCfg.sqf]: There's no need for adding another property, if `isActive` is not true the skill is passive.

File: game/functions/systems/skills/server/fn_skills_teachSkill.sqf [hunk 2]
```diff
@@ -2,7 +2,7 @@
     File: fn_skills_teachSkill.sqf
     Author: veteran29
     Date: 2023-01-27
-    Last Update: 2023-02-25
+    Last Update: 2023-05-21
```
Comment (veteran29) [inline game/functions/systems/skills/server/fn_skills_teachSkill.sqf]: The function was not changed.
```suggestion
    Last Update: 2023-02-25
```

File: game/functions/systems/skills/client/active/fn_skills_active_skillWheelActivate.sqf [hunk 3]
```diff
@@ -26,5 +26,5 @@ if (_slot call vgm_c_fnc_skills_active_isSlotOnCooldown) exitWith {
 };
 
 player call (_skill get "codeActivate");
-private _cooldownUntil = time + (_skill get "cooldown");
+private _cooldownUntil = time + (_skill get "cooldown"); // cooldown handling probably should be server-sided as well.
```
Comment (veteran29) [inline game/functions/systems/skills/client/active/fn_skills_active_skillWheelActivate.sqf]: I don't see a point in comments like this in the code. If it's an issue bring it up in the channels and let's make a task for that.

I also don't think we need server side cooldowns right now, reconnecting to skip skill cooldown is a lot of hassle with how slow arma loads.

File: game/functions/systems/skills/client/applicators/fn_skills_apply_hoofIt.sqf [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/*
+    File: fn_skills_apply_hoofIt.sqf
+    Author: Savage Game Design
+    Date: 2023-05-20
+    Last Update: 2023-05-21
+    Public: No
+
+    Description:
+        Disables stamina for 30 seconds.
+        WARNING: This is ran on the client-side.
```
Comment (veteran29) [inline game/functions/systems/skills/client/applicators/fn_skills_apply_hoofIt.sqf]: No point in this, this is in client side folder. All skill effects run client side.
```suggestion
```

File: game/functions/systems/skills/client/applicators/fn_skills_apply_steadyAim.sqf [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    Example(s):
+        call vgm_c_fnc_skills_apply_steadyAim;
+*/
+
+private _skill = ["combatTree", "steadyAim"] call vgm_g_fnc_skills_getByPath;
+private _skillTree = _skill call vgm_g_fnc_skills_getSkillTreeFromSkill;
+private _skillTreePoints = [_skillTree] call vgm_g_fnc_skills_getSkillTreePoints;
+
+private _coef = [1 - (0.01 * _skillTreePoints), 0, 1] call BIS_fnc_clamp;
```
Comment (veteran29) [inline game/functions/systems/skills/client/applicators/fn_skills_apply_steadyAim.sqf]: BI func is slow.
```suggestion
private _coef = (1 - 0.01 * _skillTreePoints) max 0 min 1;
```

File: game/functions/systems/skills/client/applicators/fn_skills_apply_steadyAim.sqf [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+player setCustomAimCoef _coef; // Probably should store this in the db for stacking purposes.
```
Comment (veteran29) [inline game/functions/systems/skills/client/applicators/fn_skills_apply_steadyAim.sqf]: Same as with other comment, also no point in storing in DB, if we would add any stacking one local function for aim modifiers will be enough.
```suggestion
player setCustomAimCoef _coef;
```

File: game/functions/systems/skills/client/applicators/fn_skills_apply_athletic.sqf [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+    Example(s):
+        call vgm_c_fnc_skills_apply_athletic;
+ */
+
+private _skill = ["combatTree", "athletic"] call vgm_g_fnc_skills_getByPath;
+private _skillTree = _skill call vgm_g_fnc_skills_getSkillTreeFromSkill;
+private _skillTreePoints = [_skillTree] call vgm_g_fnc_skills_getSkillTreePoints;
+
+private _coef = [1 - (_skillTreePoints * 0.1), 0.1, 1] call BIS_fnc_clamp;
```
Comment (veteran29) [inline game/functions/systems/skills/client/applicators/fn_skills_apply_athletic.sqf]: BI func is 10 times slower.
```suggestion
private _coef = (1 - _skillTreePoints * 0.1) max 0.1 min 1;
```

File: game/functions/systems/skills/global/fn_skills_getSkillTreePoints.sqf [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+	Last Update: 2023-05-13
+	Public: No
+	Description:
+		No description added yet.
+	Parameter(s):
+		_tree - The tree to get the skill points from. [STRING]
+	Returns:
+		Number of skills points spent in tree. [Number]
+	Example(s):
+		[_tree] call vgm_g_fnc_skills_getSkillTreePoints;
```
Comment (veteran29) [inline game/functions/systems/skills/global/fn_skills_getSkillTreePoints.sqf]: Missing description, spacing.

File: game/functions/systems/skills/server/fn_skills_forgetSkills.sqf [hunk 9]
```diff
@@ -2,7 +2,7 @@
     File: fn_skills_teachSkill.sqf
     Author: veteran29
     Date: 2023-01-27
-    Last Update: 2023-02-28
+    Last Update: 2023-05-21
```
Comment (veteran29) [inline game/functions/systems/skills/server/fn_skills_forgetSkills.sqf]: The function was not changed.
```suggestion
    Last Update: 2023-02-28
```

File: game/functions/systems/skills/server/network/fn_skills_handle_skillLearnRequest.sqf [hunk 10]
```diff
@@ -45,6 +45,8 @@ private _skillPointsSpent = _skillsData get "skillPointsSpent";
 _skillsData set ["skillPoints", _skillPoints - _skillCost];
 _skillsData set ["skillPointsSpent", _skillPointsSpent + _skillCost];
 
+[] remoteExecCall ["vgm_c_fnc_skills_reapply", _player];
```
Comment (veteran29) [inline game/functions/systems/skills/server/network/fn_skills_handle_skillLearnRequest.sqf]: I feel like it would be better to handle this client side, with event handler. See how skills that need to be re-applied on respawn are handled, `client/skills_preInit`.
```suggestion
```

File: game/functions/systems/skills/server/network/fn_skills_handle_skillRespecRequest.sqf [hunk 11]
```diff
@@ -30,6 +30,8 @@ private _skillPointsSpent = _skillsData get "skillPointsSpent";
 _skillsData set ["skillPoints", _skillPoints + _skillPointsSpent];
 _skillsData set ["skillPointsSpent", 0];
 
+[] remoteExecCall ["vgm_c_fnc_skills_reapply", _player];
```
Comment (veteran29) [inline game/functions/systems/skills/server/network/fn_skills_handle_skillRespecRequest.sqf]: This is exactly why I feel it should be in event handler. Duplication and fragmentation of logic into multiple areas.
```suggestion
```

File: game/functions/systems/skills/client/fn_skills_reapply.sqf [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        Nothing
+
+    Example(s):
+        call vgm_c_fnc_skills_reapply;
+        "primary" call vgm_c_fnc_skills_reapply;
+ */
+
+params [["_type", "passive"]];
+
+private _skillType = switch (_type) do {
```
Comment (veteran29) [inline game/functions/systems/skills/client/fn_skills_reapply.sqf]: As mentioned earlier checking if skill is not active would be enough, alternative to this is that every skill that needs an update after other skills are learned could add it's own event handler in application function. Or addition of `appllyOnLearn` flag.

**Review by veteran29 — CHANGES_REQUESTED** [review]

File: game/functions/systems/skills/client/active/fn_skills_active_skillWheelActivate.sqf (hunk 3 above)
Comment (cerebral-314) [reply game/functions/systems/skills/client/active/fn_skills_active_skillWheelActivate.sqf]: They're mostly for me actually... forgot to remove them.  Thanks

File: game/configs/mission/cfg_skills.hpp [hunk 13]
```diff
@@ -27,27 +27,37 @@ class vgm_skillTrees {
...
+                    applyOnRespawn = 1;
                 };
             };
 
             class tier_2 {
-                class skill_3: vgm_skillTemplate {};
-                class skill_4: vgm_skillTemplate {};
+                class athletic: vgm_skillTemplate {
+                    displayName = "Athletic";
+                    description = "Decreases stamina comsumption from current weight.";
```
Comment (veteran29) [inline game/configs/mission/cfg_skills.hpp]: I think we should use stringtable for "actual" skill implementations.

File: game/functions/systems/skills/client/applicators/fn_skills_apply_hoofIt.sqf [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+    Returns:
+        [BOOL] - Skill activated successfully.
+
+    Example(s):
+        call vgm_c_fnc_skills_apply_hoofIt;
+ */
+
+player enableStamina false;
+systemChat "You feel a sudden burst of energy!";
```
Comment (veteran29) [inline game/functions/systems/skills/client/applicators/fn_skills_apply_hoofIt.sqf]: Stringtable too.

**Comment by veteran29** [comment]
I think I'll move each tree skills into separate subfolders. 

`skill/actives/infantryman` etc.

File: game/functions/systems/skill/passives/fn_skill_passives_preInit.sqf [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+if (!hasInterface) exitWith {};
+
+["vgm_mission_deploy_local", {
+    params ["_mission"];
+
+    format ["Applying Rifleman/Born Leader for mission: %1", _mission] call vgm_g_fnc_logDebug;
+
+    private _group = _mission get "group";
+    _group = group player; // TODO remove after "group" in missions is merged
```
Comment (Spoffy) [inline game/functions/systems/skill/passives/fn_skill_passives_preInit.sqf]: Think we can remove this now :)

File: game/functions/systems/skill/passives/infantryman/fn_skill_passives_infantryman_overprepared.sqf [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        || {_unit ammo _weapon > 0 || {
+            private _compatibleMags = compatibleMagazines [primaryWeapon _unit, currentMuzzle _unit];
+            _compatibleMags findAny (magazines _unit) > -1
+        }}
+    ) exitWith {};
+
+    // fire only once per mission
+    private _usedInMission = _unit getVariable ["vgm_c_skill_passives_overpreparedMission", createHashMap];
+    private _currentMission = [] call vgm_c_fnc_missions_getCurrentMission;
+    if (_usedInMission isEqualTo _currentMission) exitWith {};
```
Comment (Spoffy) [inline game/functions/systems/skill/passives/infantryman/fn_skill_passives_infantryman_overprepared.sqf:40]: Is it worth using `isEqualRef` here, to avoid deep-comparing the hashmaps? Could also use the mission ID, but I think isEqualRef would be nicer.

File: game/functions/systems/skills/client/active/fn_skills_active_skillWheelActivate.sqf [hunk 17]
```diff
@@ -28,7 +28,10 @@ if (_slot call vgm_c_fnc_skills_active_isSlotOnCooldown) exitWith {
 };
 
 player call (_skill get "codeActivate");
-private _cooldownUntil = time + (_skill get "cooldown");
+
+private _cooldownTime = (_skill get "cooldown") * (player getVariable ["vgm_c_skills_cooldownCoef", 1]);
+private _cooldownUntil = time + _cooldownTime; // cooldown handling probably should be server-sided as well.
```
Comment (Spoffy) [inline game/functions/systems/skills/client/active/fn_skills_active_skillWheelActivate.sqf]: Are you thinking from a security point of view?

If so, I agree, though hackers zeroing the cooldown isn't much of a problem I think, so happy to leave this for the future. 

Could throw a 'TODO:' prefix on it if you think this is worth doing later (easier to find). 

TL;DR - I agree it's not worth doing right now.

**Review by Spoffy — CHANGES_REQUESTED** [review]

File: game/functions/systems/skills/client/active/fn_skills_active_skillWheelActivate.sqf (hunk 17 above)
Comment (veteran29) [reply game/functions/systems/skills/client/active/fn_skills_active_skillWheelActivate.sqf]: Mostly for ultimate abuse, use skill, disconnect, connect, join mission in progress. Skills available again.

File: game/functions/systems/skill/passives/infantryman/fn_skill_passives_infantryman_overprepared.sqf (hunk 16 above)
Comment (veteran29) [reply game/functions/systems/skill/passives/infantryman/fn_skill_passives_infantryman_overprepared.sqf:40]: I'm not a fan of reference checking in SQF, in case some logic in netmap handling changes it could easily break.

If the items are same reference there's zero performance benefit of using `isEqualRef`, `isEqualTo` starts with ref check.

```sqf
a = createHashMap;

for "_i" from 1 to 1e5 do {
	a set [_i, _i];
};

b = a;

// separate execution
// Execution Time: 0.0003 ms  |  Cycles: 10000/10000  |  Total Time: 3 ms
a isEqualTo b

// separate execution
// Execution Time: 0.0003 ms  |  Cycles: 10000/10000  |  Total Time: 3 ms
a isEqualRef b
```

```sqf
a = createHashMap;

for "_i" from 1 to 1e5 do {
	a set [_i, _i];
};

b = +a; // copy the hashmap

// separate execution
// Execution Time: 15.2306 ms  |  Cycles: 66/10000  |  Total Time: 1005 ms
a isEqualTo b

// separate execution
// Execution Time: 0.0003 ms  |  Cycles: 10000/10000  |  Total Time: 3 ms
a isEqualRef b;
```

Comment (Spoffy) [reply game/functions/systems/skill/passives/infantryman/fn_skill_passives_infantryman_overprepared.sqf:40]: Fair enough, that makes sense! Thanks for clarifying!
